File: C#/Heap or Priority Queue/2336. Smallest Number in Infinite Set.cs
/* 2336. Smallest Number in Infinite Set
Difficulty: Medium
Submission Link: https://leetcode.com/problems/smallest-number-in-infinite-set/submissions/1387469446/ */

public class SmallestInfiniteSet {

    PriorityQueue<int, int> heap = new ();
    HashSet<int> itemsRemove = new HashSet<int>();

    public SmallestInfiniteSet() {
       for(int i = 1; i <= 1000; i++)
       {
           heap.Enqueue(i, i);
       }
    }
    
    public int PopSmallest() {  
        int min = heap.Dequeue();
        itemsRemove.Add(min);
        return min;
    }
    
    public void AddBack(int num) {
        int min = heap.Peek();
        if(min > num)
        {
            heap.Enqueue(num, num);
            itemsRemove.Remove(min);  
        }else if(min < num)
        {
            if(itemsRemove.Contains(num))
            {
                heap.Enqueue(num, num);
                itemsRemove.Remove(min);
            }
        }
    }
}

File: Python3/Heap or Priority Queue/2336. Smallest Number in Infinite Set.py
""" 2336. Smallest Number in Infinite Set
Difficulty: Medium
Submission Link: https://leetcode.com/problems/smallest-number-in-infinite-set/submissions/1387468961/ """

class SmallestInfiniteSet:
    def __init__(self):
        self.set = set(range(1, 1001))

    def popSmallest(self) -> int:
        if not self.set:
            return None

        min_val = min(self.set)
        self.set.remove(min_val)
        return min_val

    def addBack(self, num: int) -> None:
        self.set.add(num)